### ISSUE TITLE
Recursive definitions and invariant

### ISSUE THREAD

**Issue by jhjourdan** [xldenis/creusot#879]
```rust
extern crate creusot_contracts;
use creusot_contracts::{*, invariant::{Invariant, inv}};

struct S { }

impl Invariant for S {
    #[predicate]
    #[open]
    fn invariant(self) -> bool {
        !inv(self)
    }
}
```

This leads to an absurd axiom.

Basically, we need to check that inv is not used (directly or not) in the definition of `invariant`.

Note that, for this reason, Why3 refuses recursive types with invariants:

```
use option.Option

type t = {
    x : option t
}
invariant { not (x = None) }
```

is refused by Why3.

Also, Why3 rejects the following:

```
use option.Option

type t 'a = {
    x : option 'a
}
invariant { not (x = None) }

type u = { y : t u }
```

By telling that u is not used in a strictly positive fashion in its definition (i.e., a type with an invariant has no strictly positive type parameters).

Depending on the solution we find for this bug, we may need to take into account those problems in Creusot too.

**Comment by voidc** [comment]
Good catch! The problem also exists with an invariant like `forall<x: S> false`. I'm not sure what the best way to fix this is.

**Comment by xldenis** [comment]
The solution @jhjourdan has proposed is to remove the implicit invariant assumption in quantifiers.

**Comment by dianegolfouse** [comment]
Here are some ideas we had for solving the issue in the termination check:

`inv` (alternatively, every function that calls `inv`) gets an arc to every implementation of `Invariant::invariant`.
 
This prevents implementations of `Invariant` from calling `inv`. This is mainly problematic for types like `Seq`, that need to axiomatize their own structural invariant.
We could then have an exception for such types:
- either we introduce a bespoke attribute, or reuse `trusted_ignore_structural_inv`... This is very ad-hoc
- Or we could find a criteria for calling `inv` in `invariant` implementations: @jhjourdan's idea on this was that
  - an implementation of `Invariant` must not introduce bounds on the type parameters
  - `inv` can only be called on type variables introduced on the impl block. This means there would automatically be a structural variant decreasing, so this is ok.
  
  (am I correctly relating what you said?)

Note that this particular problem is not, in fact, limited to `inv`: it also exists for `resolve`, problably `precondition`/`postcondition`, and maybe others. This is an issue with creusot builtins.
